src/index.js

```diff
@@ -1,6 +1,20 @@
 const fibonacci = (n) => {
+  const data = [];
 
+  for (let index = 0; index < n; index++) {
```

Comment:
Te propongo intentar con recursividad y programación dinámica. Adapta la idea que David hizo en python a javascript de la case https://platzi.com/clases/1835-programacion-estocastica/26430-optimizacion-de-fibonacci/

Reply:
Claro que si muchas gracias voy a verlo y realizarlo para aprender otra forma de resolverlo... Gracias por todos tus comentarios.

Reply:
ya lo hice, si entendi bien, la idea es reutilizar las iteraciones para no volver a crear el array y no realizar el recorrido con el for para no utilizar recursos innecesarios y que sea mas rapida la ejecución.

Por eso hice lo siguiente:
1. El objeto data(array que almacena los números de fibonacci) se declaró por fuera de la función para no resetearlo.
2. realice las validaciones si el objeto ya posee la cantidad de números que está solicitando solo consulta el array y trae los datos hasta esa posición (en este caso la variable n recibida por la función) . 
3. si el número que se solicita no existe va y realiza el recorrido desde la última posición existente hasta el número que fue solicitado.

quiero saber si tu eres tan amable de confirmarme si el concepto que tengo de recursividad es el correcto.

Gracias por todo.

Reply:
Recursividad se refiere a que se llama a si mismo.

El ejemplo mas sencillo es el factorial, por  ejemplo, `5! = 5 * 4 * 3 * 2 * 1 = 5 * 4!`. En general `(n+1)! = (n + 1) * n!`.

En el caso de la serie de fibonacci, su definición recursiva es

- `fibonacci(1) = 1`
- `fibonacci(2) = 2`
- para toda `n > 2`,  `fibonacci(n) = fibonacci(n-2) + fibonacci(n-1)`

Esto en código se podría ver algo así

```js
const fibonacci = n => (n <= 2) ? 1 : fibonacci(n-1) + fibonacci(n-2);
```

Sin embargo esto genera un problema de cálculos, ya que si crece mucho el número `n` se tarda mucho.

Si analizamos el ciclo de ejecución de fibonacci(4) verás que se hacen muchos cálculos iguales:

1. Solicitamos fibonacci(5)
2. 5 <= 2 es falso, por lo que va a regresar fibonacci(4) + fibonacci(3). Para poder responder el valor necesitamos calcular fibonacci(4) y fibonacci(3), vamos primero por fibonacci(4).
3. Solicitamos fibonacci(4)
4. 4 <= 2 es falso por lo que va a regresar fibonacci(3) + fibonacci(2). Para poder responder el valor necesitamos calcular fibonacci(3) y fibonacci(2). Vamos primero por fibonacci(3);
5. Solicitamos fibonacci(3)
6. 3 <= 2 es falso por lo que vamos a regresar fibonacci(2) + fibonacci(1). Para poder responder el valor necesitamos calcular fibonacci(2) y fibonacci(1). Veamos primero fibonacci(2).
7. Solicitamos fibonacci(2).
8. 2 <= 2 es verdadero, por lo que regresamos 1.
9. Ya tenemos el valor de fibonacci(2) = 1, que se requería en el paso 6. Ahora requerimos fibonacci(1).
10. Solicitamos fibonacci(1).
11. 1 <= 2 es verdadero, por lo que, regresamos 1.
12. Ya tememos el valor de fibonacci(2) = 1 y fibonacci(1) = 1, por lo que fibonacci(2) + fibonacci(1) = 2. Podemos responder a la solicitud en 5 con fibonacci(3) = 2.
13. Ya tememos fibonacci(3) = 2, del paso 4, ahora requerimos fibonacci(2)
14. Repetimos los paso del 7-8.
15. Ya tenemos que fibonacci(3) = 2 y fibonacci(2) = 1, que solicitó el paso 4. Por lo que fibonacci(3) + fibonacci(2) = 3. Y podemos contestar a la solicitud del paso 3.
16. Ya tenemos que fibonacci(4) = 3, que se solicitó en el paso 2. Ahora hay que calcular fibonacci(3)
17. Repetimos los pasos del 5 al 12.
18. Ya tenemos el valor de fibonacci(4) = 3, fibonacci(3)= 2, por lo que fibonacci(4) + fibonacci(3) = 5.
19. Retornamos que fibonacci(5) es igual a 5.

Así es como funciona la recusividad. En los paso 14 y 17, están condensados, por que se volvería a hacer los cálculos. Si se pudieran guardar los fibonacci ya calculados, podríamos responder enseguida y no volverlos a solicitar. Este algoritmo crece del orden O(n^2), pero si le aplicas memoria crece al orden O(n). Por lo que la solución óptima recusiva requiere memoria.